docs/design/secure-config.md

```diff
@@ -0,0 +1,347 @@
+  available.
+
+#### Distinguishing copies from moves
+
+A user's expectation is that if they run `cp -r old_repo new_repo`, then modify
+the old repo's config, the new repo's config is not affected. Thus, we need to
+make sure that the repo remains in a 1:1 relationship with the config. Multiple
+repos should not point to the same config.
+
+To achieve this, we point the repo at the config, and the config back at the
+repo. If the path stored in the config doesn't match, we know that something has
+happened. To decide precisely what happened:
+* If the path stored in the config no longer exists, we assume it's a move
+  * We update `metadata.path` to point to the new path
+* Otherwise, we attempt to write to a temporary file in `$NEW_REPO/.jj/repo`.
+  * If it shows up in the original repository, it's some kind of reference
+    (link, mount, etc.), so we don't do anything.
+  * Otherwise, it's a copy, so we generate a new config ID and copy the old
+    config to it
+
+Note that with a copy, we only actually copy the config when the user runs a
+`jj` command. This means that you can end up in a situation where you:
+1. `cp -r old_repo new_repo`
+2. In `old_repo`, `jj config set --repo`
+3. Run a jj command in the copy. This copies the config, including the config
+   from step 2.
+
+However, this is inherent to storing config out of the repo and is thus
+unavoidable.
+
+### Garbage collection
+
+We could, in the future, add a `gc` command to garbage-collect configs to
+deleted repo configs. However, there are some things to consider before doing
+so:
+* Each config would likely be very small, so cleaning it up may have limited
+  benefit.
+* It is impossible to distinguish "deleted" from "moved".
+* If you have something like a chroot or a dual boot where you share the
+  config, you may have references to config IDs with a different path.
+
+### Attack vectors remaining
+
+Unfortunately, there is no way to distinguish copying / moving from a replay
+attack. The attacker, if they know a config ID that exists on your system, can
+create a repo with the same config ID. However, the fact that the config itself is
+stored out of repo inherently prevents simple replay attacks. In order for the
+attacker to exploit this, they would need to:
+
+*   Know your config ID (requires uploading a zip file or something similar)
+*   Get lucky by the victim having a “risky” per-repo config
+*   Eg. fix.tools pointing to `$repo/formatter`
+*   Know how to exploit it
+*   Because your config file is stored out-of-repo, the attacker will likely not
+    know any of this without some social engineering
+
+We intentionally choose not to deal with this kind of attack in the initial
+version, and have no current intention to solve it in future versions either (as
+the UX impact would likely be much larger than any benefit to security).
+
+We could potentially add this as an opt-in feature in the future, but it has
```

Comment:
What does "this" mean here? I'm guessing it's some way of addressing this attack but it also seems like there's some particular solution you have in mind since you're talking about the UX impact. I think it would be good to explain what solution you have in mind. Alternatively, you can be less specific here and leave it for us to think more about in the future.

Reply:
I don't have a particular solution in mind - rather, I'm trying to say that any solution to solve this will inherently get false positives, and so any attempt to prevent this attack vector will also prevent copies / moves, which is the UX cost. I've elaborated in the doc.